main/src/ca/uwaterloo/flix/language/phase/MonomorphEnums.scala

```diff
@@ -0,0 +1,449 @@
+      Expression.Unary(sop, visitExp(exp), visitType(tpe), visitType(pur), loc)
+    case Expression.Binary(sop, exp1, exp2, tpe, pur, loc) =>
+      Expression.Binary(sop, visitExp(exp1), visitExp(exp2), visitType(tpe), visitType(pur), loc)
+    case Expression.Let(sym, mod, exp1, exp2, tpe, pur, loc) =>
+      Expression.Let(sym, mod, visitExp(exp1), visitExp(exp2), visitType(tpe), visitType(pur), loc)
+    case Expression.LetRec(sym, mod, exp1, exp2, tpe, pur, loc) =>
+      Expression.LetRec(sym, mod, visitExp(exp1), visitExp(exp2), visitType(tpe), visitType(pur), loc)
+    case Expression.Region(tpe, loc) =>
+      Expression.Region(visitType(tpe), loc)
+    case Expression.Scope(sym, regionVar, exp, tpe, pur, loc) =>
+      // The region variable has been rendered redundant by Monomorph.
+      // It has replaced the region with pure/impure and the variable could
+      // conceptually be removed.
+      Expression.Scope(sym, regionVar, visitExp(exp), visitType(tpe), visitType(pur), loc)
+    case Expression.ScopeExit(exp1, exp2, tpe, pur, loc) =>
+      Expression.ScopeExit(visitExp(exp1), visitExp(exp2), visitType(tpe), visitType(pur), loc)
+    case Expression.IfThenElse(exp1, exp2, exp3, tpe, pur, loc) =>
+      Expression.IfThenElse(visitExp(exp1), visitExp(exp2), visitExp(exp3), visitType(tpe), visitType(pur), loc)
+    case Expression.Stm(exp1, exp2, tpe, pur, loc) =>
+      Expression.Stm(visitExp(exp1), visitExp(exp2), visitType(tpe), visitType(pur), loc)
+    case Expression.Discard(exp, pur, loc) =>
+      Expression.Discard(visitExp(exp), visitType(pur), loc)
+    case Expression.Match(exp, rules, tpe, pur, loc) =>
+      val newRules = rules.map {
+        case LoweredAst.MatchRule(pat, guard, exp) =>
+          LoweredAst.MatchRule(visitPat(pat), guard.map(visitExp), visitExp(exp))
+      }
+      Expression.Match(visitExp(exp), newRules, visitType(tpe), visitType(pur), loc)
+    case Expression.TypeMatch(exp, rules, tpe, pur, loc) =>
+      val newRules = rules.map {
+        case LoweredAst.TypeMatchRule(sym, tpe, exp) =>
+          LoweredAst.TypeMatchRule(sym, visitType(tpe), visitExp(exp))
+      }
+      Expression.TypeMatch(visitExp(exp), newRules, visitType(tpe), visitType(pur), loc)
+    case Expression.RelationalChoose(exps, rules, tpe, pur, loc) =>
+      val newRules = rules.map {
+        case LoweredAst.RelationalChoiceRule(pat, exp) =>
+          val newPat = pat.map {
+            case RelationalChoicePattern.Wild(loc) => RelationalChoicePattern.Wild(loc)
+            case RelationalChoicePattern.Absent(loc) => RelationalChoicePattern.Absent(loc)
+            case RelationalChoicePattern.Present(sym, tpe, loc) => RelationalChoicePattern.Present(sym, visitType(tpe), loc)
+          }
+          LoweredAst.RelationalChoiceRule(newPat, visitExp(exp))
+      }
+      Expression.RelationalChoose(exps.map(visitExp), newRules, visitType(tpe), visitType(pur), loc)
+    case Expression.Tag(sym, exp, tpe, pur, loc) =>
```

Comment:
It might be good to add some comments to call out that this is where things happen :)